main.tf

```diff
@@ -38,47 +50,96 @@ resource "aws_efs_access_point" "default" {
 
   file_system_id = join("", aws_efs_file_system.default.*.id)
 
-  posix_user {
-    gid = var.access_points[each.key]["posix_user"]["gid"]
-    uid = var.access_points[each.key]["posix_user"]["uid"]
-    # Just returning null in the lookup function gives type errors and is not omitting the parameter, this work around ensures null is returned.
-    secondary_gids = lookup(lookup(var.access_points[each.key], "posix_user", {}), "secondary_gids", null) == null ? null : null
+  dynamic "posix_user" {
+    for_each = local.secondary_gids[each.key] != null ? ["true"] : []
+
+    content {
+      gid            = var.access_points[each.key]["posix_user"]["gid"]
+      uid            = var.access_points[each.key]["posix_user"]["uid"]
+      secondary_gids = local.secondary_gids[each.key] != null ? split(",", local.secondary_gids[each.key]) : null
+    }
```

Comment:
I am not following this. `var.access_points` needs better documentation, but it looks to me like this is wrong, because either it makes `gid` and `uid` ignored if there are no secondary GIDs provided or it always provisions `gid` and `uid` in which case the `posix_user` block does not need to be dynamic. If the code is behaving correctly, then `local.secondary_gids` is a bad name for that variable.

Reply:
Good catch. I replaced the `for_each` value to use posix user and updated the description of `var.access_points`.